### PR TITLE
Fixed #547 : Email subscription settings for individual bouts

### LINKED ISSUE #547: Email subscription settings for individual bouts
At the moment, any user/member of Netbout will receive an email for any message that arrives at any bout. It would be great if we could follow/subscribe to individual bouts. Emails should be sent only for new messages being posted into the subscribed ones.

### PR DESCRIPTION
**Pull request by erimerturk**
Implemented Email subscription settings for individual bout #547

### CODE REVIEW

**Comment by karato** [comment]
@erimerturk Thanks, I will find someone to review this PR soon

**Comment by karato** [comment]
@pinaf do this review

**Comment by pinaf** [comment]
@erimerturk hmm I don't see how this solves the original issue. You added a property to the bout, but the property (subscribed or not) should belong to the pair (bout,user). Or did I misunderstand your solution?

**Comment by erimerturk** [comment]
@pinaf actually i added that attribute to dy-friends table. which has bout and user pair. before sending email <b>EmMessages </b> should filter friend by subscription.

File: netbout-web/src/main/java/com/netbout/email/EmMessages.java [hunk 1]
```diff
@@ -93,6 +93,13 @@
         this.self = slf;
     }
 
+    /**
+     * Post a new message.
```
Comment (pinaf) [inline netbout-web/src/main/java/com/netbout/email/EmMessages.java]: @erimerturk no need for javadoc here. for the puzzle, use multiple `//` comments

File: netbout-web/src/test/java/com/netbout/dynamo/DyBoutITCase.java [hunk 2]
```diff
@@ -61,4 +62,20 @@ public void renamesBout() throws Exception {
...
+    @Test
+    public void changesSubscriptionBout() throws Exception {
+        final String alias = "maxi";
+        final Aliases aliases =
+            new DyBase().user(new URN("urn:test:1890")).aliases();
+        aliases.add(alias);
+        final Inbox inbox = aliases.iterate().iterator().next().inbox();
+        final Bout bout = inbox.bout(inbox.start());
+        bout.subscribe(false);
+        Assert.assertFalse(bout.subscription());
```
Comment (pinaf) [inline netbout-web/src/test/java/com/netbout/dynamo/DyBoutITCase.java]: @erimerturk use `MatcherAssert.assertThat`

File: netbout-web/src/main/java/com/netbout/dynamo/DyBout.java [hunk 3]
```diff
@@ -121,6 +121,21 @@ public void rename(final String text) throws IOException {
...
+            this.item.get(DyFriends.ATTR_SUBSCRIPTION).getS()
+        );
+    }
+
+    @Override
+    public void subscribe(final boolean subs) throws IOException {
+        this.item.put(
+            new AttributeUpdates().with(DyFriends.ATTR_SUBSCRIPTION, subs)
+        );
+        Logger.info(this, "bout #%d subscribe to \"%s\"", this.number(), subs);
```
Comment (pinaf) [inline netbout-web/src/main/java/com/netbout/dynamo/DyBout.java]: @erimerturk let's add info on who subscribed to the bout

File: netbout-client/src/main/java/com/netbout/mock/H2Sql.java [hunk 4]
```diff
@@ -66,7 +66,7 @@
         final String[] stmts = {
             // @checkstyle LineLength (5 lines)
             "CREATE TABLE alias (name VARCHAR, urn VARCHAR, photo VARCHAR, locale VARCHAR, email VARCHAR)",
-            "CREATE TABLE bout (number BIGINT AUTO_INCREMENT, title VARCHAR, date TIMESTAMP DEFAULT CURRENT_TIMESTAMP)",
+            "CREATE TABLE bout (number BIGINT AUTO_INCREMENT, title VARCHAR, date TIMESTAMP DEFAULT CURRENT_TIMESTAMP, subscription INTEGER)",
```
Comment (pinaf) [inline netbout-client/src/main/java/com/netbout/mock/H2Sql.java]: @erimerturk this new `subscription` flag - is it related to a specific user?

**Comment by pinaf** [comment]
@erimerturk 3 comments and 1 question above

File: netbout-client/src/main/java/com/netbout/mock/H2Sql.java (hunk 4 above)
Comment (erimerturk) [inline netbout-client/src/main/java/com/netbout/mock/H2Sql.java]: @pinaf u r right it is my mistake. it should be defined on friend table. i fixed it

**Comment by pinaf** [comment]
@erimerturk excellent!

**Comment by pinaf** [comment]
@rultor merge

**Comment by erimerturk** [comment]
@pinaf thank you!

File: netbout-web/src/main/java/com/netbout/dynamo/DyBout.java [hunk 5]
```diff
@@ -121,6 +121,26 @@ public void rename(final String text) throws IOException {
     }
 
     @Override
+    public boolean subscription() throws IOException {
+        return Boolean.parseBoolean(
+            this.item.get(DyFriends.ATTR_SUBSCRIPTION).getS()
```
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/dynamo/DyBout.java]: @erimerturk what happens here for existing records, which hasn't the subscription attribute before ?
returns null or what? if so let's add default value `true`

**Comment by dmzaytsev** [comment]
@erimerturk one comment above

File: netbout-web/src/main/java/com/netbout/dynamo/DyBout.java (hunk 5 above)
Comment (erimerturk) [inline netbout-web/src/main/java/com/netbout/dynamo/DyBout.java]: @dmzaytsev u r right, i will fix it

**Comment by erimerturk** [comment]
@dmzaytsev please see changes, thank you

File: netbout-web/src/main/java/com/netbout/dynamo/DyBout.java [hunk 6]
```diff
@@ -121,6 +121,30 @@ public void rename(final String text) throws IOException {
     }
 
     @Override
+    @SuppressWarnings("PMD.OnlyOneReturn")
```
Comment (dmzaytsev) [inline netbout-web/src/main/java/com/netbout/dynamo/DyBout.java]: @erimerturk let's remove PMD suppression here
we can use  `( ? : ):` operator. isn't it ?
or create a variable for the return value

**Comment by pinaf** [comment]
@dmzaytsev qulice complains about the ternary operator. you could add a boolean variable that starts with a value and only changes inside the if.

**Comment by erimerturk** [comment]
@dmzaytsev @pinaf ready for review, removed OnlyOneReturn suppress, thank you

**Comment by pinaf** [comment]
@rultor merge

**Comment by dmzaytsev** [comment]
@rultor merge

@pinaf @erimerturk thanks!

**Comment by dmzaytsev** [comment]
@rultor deploy

**Comment by karato** [comment]
@ypshenychka review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by karato** [comment]
@rultor deploy

**Comment by ypshenychka** [comment]
@karato Quality is good here.

**Comment by karato** [comment]
> @karato Quality is good here.

@ypshenychka thanks, got it, great to see high-quality tickets here

**Comment by karato** [comment]
@pinaf I added 10 mins to @ypshenychka (for QA review) in transaction 59055119

since quality is good, I just added 10 mins to @dmzaytsev (our architect) in transaction 59055132

thanks, paid, **21 mins** to your account, payment ID is `AP-7SK16149GM521604F`, task took 95 hours and 24 mins to complete

review comments (c=6) added as a bonus

+21 added to your rating, current score is: [+6602](http://www.netbout.com/b/35128?open=rating)

**Pull request merged**